symengine/polys/basic_conversions.h

```diff
@@ -0,0 +1,221 @@
+#include <symengine/visitor.h>
+
+namespace SymEngine
+{
+
+template <typename P>
+RCP<const P> from_basic(const RCP<const Basic> &basic,
+                        const RCP<const Basic> &gen, bool ex = false);
+template <typename P>
+RCP<const P> from_basic(const RCP<const Basic> &basic, bool ex = false);
+
+template <typename T, typename P>
+enable_if_t<std::is_same<T, UExprDict>::value, T>
+_basic_to_upoly(const RCP<const Basic> &basic, const RCP<const Basic> &gen);
+
+template <typename T, typename P>
+enable_if_t<std::is_base_of<UIntPolyBase<T, P>, P>::value, T>
+_basic_to_upoly(const RCP<const Basic> &basic, const RCP<const Basic> &gen);
+
+template <typename P, typename V>
+class BasicToUPolyBase : public BaseVisitor<V>
+{
+public:
+    RCP<const Basic> gen;
+    using D = typename P::container_type;
+    D dict;
+
+    D apply(const Basic &b, const RCP<const Basic> &gen_)
+    {
+        gen = gen_;
+        b.accept(*this);
+        return dict;
+    }
+
+    void d_add_term(unsigned int pow, const Basic &x)
+    {
+        static_cast<V *>(this)->d_add_term(pow, x);
+    }
+
+    void bvisit(const Pow &x)
+    {
+        if (is_a<const Integer>(*x.get_exp())) {
+            int i = rcp_static_cast<const Integer>(x.get_exp())->as_int();
+            if (i > 0) {
+                dict
+                    = pow_upoly(*P::from_container(gen, _basic_to_upoly<D, P>(
+                                                            x.get_base(), gen)),
+                                i)
+                          ->get_poly();
+                return;
+            }
+        }
+
+        RCP<const Basic> genbase = gen, genpow = one, coef = one, tmp;
+        if (is_a<const Pow>(*gen)) {
+            genbase = rcp_static_cast<const Pow>(gen)->get_base();
```

Comment:
Yes. `gen` is already an RCP, but `rcp_static_cast<const Pow>` would create a new `RCP` which we don't need.
`genbase = static_cast<const Pow &>(*gen).get_base()` would work.